PR4: add explicit function return type

PPpro opened the pull request:
Re: https://github.com/cocos/3d-tasks/issues/16413

### Changelog

* add explicit function return type
* enable eslint rule `@typescript-eslint/explicit-function-return-type`, rely on PR: https://github.com/cocos/cocos-engine/pull/15311 https://github.com/cocos/cocos-engine/pull/15312 https://github.com/cocos/cocos-engine/pull/15313
* disable eslint rule `@typescript-eslint/no-inferrable-types`， this rule saying that the code below is incorrect
```ts
class A {
   a: number = 1;
}
```

### remainning issue
- https://github.com/cocos/3d-tasks/issues/16802
- https://github.com/cocos/3d-tasks/issues/16804

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!
- [ ] Your pull request title is using English, it's precise and appropriate.
- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- [ ] Document new code with comments in source code based on API docs
- [ ] Make sure any runtime log information in `log` , `error` or `new Error('')` has been moved into `EngineErrorMap.md` with an ID, and use `logID(id)` or `new Error(getError(id))` instead.
- To official teams:
  - [ ] Check that your PR is following our [guides](https://github.com/cocos/3d-tasks/blob/master/workflows/readme.md)
-->

PPpro committed:
PR4: add explicit function return type

minggo commented on cocos/render-scene/core/program-lib.ts:
```diff
@@ -565,7 +565,7 @@ export class ProgramLib {
 }
 
-export function getDeviceShaderVersion (device: Device) {
+export function getDeviceShaderVersion (device: Device): "glsl1" | "glsl3" | "glsl4" {
```
Return string instead?

minggo commented on cocos/render-scene/core/program-utils.ts:
```diff
@@ -173,12 +173,12 @@ export function populateMacros (tmpl: IProgramInfo) {
+            def._map = (value: any): number => Math.max(0, def.options!.findIndex((s): boolean => s === value));
         } else if (def.type === 'boolean') {
-            def._map = (value: any) => (value ? 1 : 0);
+            def._map = (value: any): 0 | 1 => (value ? 1 : 0);
```
I think return number is better.

minggo commented on cocos/rendering/custom/define.ts:
```diff
@@ -56,19 +56,19 @@ export enum AntiAliasing {
 }
 
-export function getRTFormatBeforeToneMapping (ppl: BasicPipeline) {
+export function getRTFormatBeforeToneMapping (ppl: BasicPipeline): Format.RGBA8 | Format.RGBA16F {
```
I think should return `Format`.

PPpro replied:
updated

minggo commented on cocos/rendering/custom/graph.ts:
```diff
@@ -156,7 +156,7 @@ export class OutEPI implements IterableIterator<EPD> {
         this.source = source;
     }
-    [Symbol.iterator] () {
+    [Symbol.iterator] (): this {
```
`this` is a valid type?

PPpro replied:
it's valid type in TS, but it's better to mark it as the typeof this

minggo commented on cocos/rendering/post-process/passes/base-pass.ts:
```diff
@@ -11,11 +11,11 @@ import { cclegacy, macro } from '../../../core';
 let _BasePassID = 0;
 
-export function getRTFormatBeforeToneMapping (ppl: BasicPipeline) {
+export function getRTFormatBeforeToneMapping (ppl: BasicPipeline): Format.RGBA8 | Format.RGBA16F {
```
I think should return Format.

minggo commented on cocos/rendering/post-process/utils/pass-context.ts:
```diff
@@ -37,35 +37,35 @@ export class PassContext {
     postProcess: PostProcess | undefined;
 
-    setClearFlag (clearFlag: ClearFlagBit) {
+    setClearFlag (clearFlag: ClearFlagBit): this {
```
Is this a valid type?

PPpro replied:
updated

minggo commented on cocos/rendering/render-stage.ts:
```diff
@@ -137,14 +137,14 @@ export abstract class RenderStage {
      * @param view The render view
      */
-    public abstract render (camera: Camera);
+    public abstract render (camera: Camera): any;
```
I think its return type is void.

minggo commented on cocos/rendering/render-stage.ts:
```diff
@@ -137,14 +137,14 @@ export abstract class RenderStage {
      * @zh 销毁函数。
      */
-    public abstract destroy ();
+    public abstract destroy (): any;
```
I think its return type is void.

minggo commented on cocos/rendering/scene-culling.ts:
```diff
@@ -37,9 +37,9 @@ const _sphere = geometry.Sphere.create(0, 0, 0, 1);
 const _tmpBoundingBox = new AABB();
 
-const roPool = new Pool<IRenderObject>(() => ({ model: null!, depth: 0 }), 128);
+const roPool = new Pool<IRenderObject>((): { model: any; depth: number; } => ({ model: null!, depth: 0 }), 128);
```
Should change to `IRenderObject`?

minggo commented on cocos/scene-graph/component-event-handler.ts:
```diff
@@ -174,18 +174,18 @@ export class EventHandler {
     }
 
-    private _compName2Id (compName) {
+    private _compName2Id (compName): any {
```
Return string?

PPpro replied:
here we use legacyCC in this implementation, so TS can only infer the return type as `any`

minggo commented on cocos/scene-graph/component-event-handler.ts:
```diff
@@ -174,18 +174,18 @@ export class EventHandler {
     }
 
-    private _compId2Name (compId) {
+    private _compId2Name (compId): any {
```
The same as above.

PPpro replied:
here we use legacyCC in this implementation, so TS can only infer the return type as any

minggo replied:
But we know its return type. Can not change to `string`?

PPpro replied:
then the TS language service would emit an error saying that `Unsafe return of an 'any' typed value`

minggo commented on cocos/serialization/ccon.ts:
```diff
@@ -36,11 +36,11 @@ export class CCON {
     }
 
-    get document () {
+    get document (): unknown {
```
Is `unknown` a valid type?

PPpro replied:
yes, it's type safer then any

minggo commented on cocos/serialization/ccon.ts:
```diff
@@ -54,15 +54,18 @@ interface CCONPreface {
 }
 
-export function encodeCCONJson (ccon: CCON, chunkURLs: string[]) {
+export function encodeCCONJson (ccon: CCON, chunkURLs: string[]): unknown {
```
The same as above.

PPpro replied:
yes

minggo commented on cocos/tween/actions/action-interval.ts:
```diff
@@ -167,13 +167,13 @@ export class ActionInterval extends FiniteTimeAction {
     }
 
-    reverse () {
+    reverse (): this {
```
Is `this` a valid type?

PPpro replied:
it's valid type in TS, but it's better to mark it as the typeof this

minggo commented on cocos/tween/tween.ts:
```diff
@@ -70,7 +70,7 @@ export class Tween<T> {
      * @param tag @en The tag set for this tween @zh 为当前缓动设置的标签
      */
-    tag (tag: number) {
+    tag (tag: number): this {
```
The same as above.

minggo replied:
Its type changes all the time?

minggo commented on cocos/ui/page-view.ts:
```diff
@@ -731,7 +731,7 @@ export class PageView extends ScrollView {
     }
 
-    protected _getDragDirection (moveOffset: Vec2) {
+    protected _getDragDirection (moveOffset: Vec2): 0 | 1 | -1 {
```
Should return number.

PPpro committed:
fix

# Conflicts:
#	cocos/2d/renderer/stencil-manager.ts
#	cocos/animation/marionette/animation-graph.ts
#	cocos/animation/marionette/event/event-binding.ts
#	cocos/animation/marionette/state-machine/motion-state.ts
#	cocos/animation/tracks/track.ts
#	cocos/animation/tracks/vector-track.ts
#	cocos/asset/assets/image-asset.ts
#	cocos/core/math/bits.ts
#	cocos/core/math/color.ts
#	cocos/core/math/mat3.ts
#	cocos/core/math/mat4.ts
#	cocos/core/math/quat.ts
#	cocos/core/math/rect.ts
#	cocos/core/math/size.ts
#	cocos/core/math/vec2.ts
#	cocos/core/math/vec3.ts
#	cocos/core/math/vec4.ts
#	cocos/core/system.ts
#	cocos/core/utils/path.ts
#	cocos/dragon-bones/CCArmatureDisplay.ts
#	cocos/gfx/base/define.ts
#	cocos/gfx/base/pipeline-sub-state.ts
#	cocos/particle-2d/png-reader.ts
#	cocos/particle-2d/tiff-reader.ts
#	cocos/particle/animator/curve-range.ts
#	cocos/particle/animator/gradient-range.ts
#	cocos/physics/cocos/builtin-rigid-body.ts
#	cocos/physics/cocos/builtin-world.ts

PPpro committed:
2d module

2d

PPpro committed:
particle module

PPpro committed:
physics module

PPpro committed:
3d module

3d

PPpro committed:
rendering module

PPpro committed:
animation module

PPpro committed:
asset module

PPpro committed:
misc && other modules

PPpro committed:
core module

PPpro committed:
fix build native ts engine && add eslint rule

minggo commented on cocos/dragon-bones/ArmatureDisplay.ts:
```diff
@@ -570,7 +570,7 @@ export class ArmatureDisplay extends UIRenderer {
 
     protected _drawIdx = 0;
-    protected _drawList = new RecyclePool<ArmatureDisplayDrawData>(() => ({
+    protected _drawList = new RecyclePool<ArmatureDisplayDrawData>((): { material: any; texture: any; indexOffset: number; indexCount: number; } => ({
```
Use ArmatureDisplayDrawData instead.

PPpro replied:
updated https://github.com/cocos/cocos-engine/pull/15314/commits/4a30db77036fbde33de0e25d66807936e3a8b685

minggo commented on cocos/dragon-bones/DragonBonesAsset.ts:
```diff
@@ -152,7 +152,7 @@ export class DragonBonesAsset extends Asset {
      * @engineInternal Since v3.7.2, this is an engine private function.
      */
-    public getAnimsEnum (armatureName: string) {
+    public getAnimsEnum (armatureName: string): { '<None>': number; } | null {
```
Please make sure it is a correct type.

PPpro replied:
yes, `{ '<None>': number; }` is an infered type from `return Enum({ '<None>': 0 })` statement

minggo commented on cocos/physics-2d/framework/physics-system.ts:
```diff
@@ -182,7 +183,7 @@ export class PhysicsSystem2D extends Eventify(System) {
      * 获取预定义的物理分组。
      */
-    public static get PhysicsGroup () {
+    public static get PhysicsGroup (): typeof PhysicsGroup {
```
What's the type of PhysicsGroup?

PPpro replied:
`PhysicsGroup` is an enum type, 

this is a trick on TS, if a variable is declared as `typeof PhysicsGroup`, that means this variable could be assigned with `PhysicsGroup` just like 
```ts
class A {
    PG: typeof PhysicsGroup  = PhysicsGroup;
}
```

we don't even need to know what's the type of `typeof PhysicsGroup`, 

This trick can help us implement the internal enum type of the class in TS

further read about Typeof Type operator in TS doc: https://www.typescriptlang.org/docs/handbook/2/typeof-types.html

minggo commented on cocos/physics/framework/physics-system.ts:
```diff
@@ -78,7 +78,7 @@ export class PhysicsSystem extends System implements IWorldInitData {
      * 获取预定义的物理分组。
      */
-    public static get PhysicsGroup () {
+    public static get PhysicsGroup (): typeof PhysicsGroup {
```
The same as above.

minggo commented on cocos/animation/core/pose.ts:
```diff
@@ -141,66 +141,66 @@ function copyTransformsWithFilter (target: TransformArray, source: Readonly<Tran
 }
 
-const blendIntoTransformArrayAt = (() => {
+const blendIntoTransformArrayAt = ((): (target: TransformArray, source: Readonly<TransformArray>, alpha: number, transformIndex: number) => void => {
```
`((): (target: TransformArray, source: Readonly<TransformArray>, alpha: number, transformIndex: number) => void => {`  ->. `((target: TransformArray, source: Readonly<TransformArray>, alpha: number, transformIndex: number) => void {`

minggo replied:
Please check other functions too.

PPpro replied:
There is nothing wrong with the original way of implementation, I think it is just more difficult to read, it may be easier to understand if it is changed to this:
```ts
type BlendIntoTransformArrayAtFunc = (target: TransformArray, source: Readonly<TransformArray>, alpha: number, transformIndex: number) => void;
const blendIntoTransformArrayAt = ((): BlendIntoTransformArrayAtFunc => {
//....
})();
```

I'll optimize this

PPpro replied:
updated here https://github.com/cocos/cocos-engine/pull/15314/commits/4a30db77036fbde33de0e25d66807936e3a8b685

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts:
```diff
@@ -257,7 +263,7 @@ const {
 
 if (EDITOR) {
-    PoseNodeApplyTransform.prototype.getTitle = function getTitle (this: PoseNodeApplyTransform) {
+    PoseNodeApplyTransform.prototype.getTitle = function getTitle (this: PoseNodeApplyTransform): string | [string, Record<string, string>] | undefined {
```
It seems returns wrong type.

PPpro replied:
no, if you check out the original implementation of `getTitle` method, it does be declared as `string | [string, Record<string, string>] | undefined`, though it's wierd to do like this.

https://github.com/cocos/cocos-engine/blob/bdcde2acf7917ea7e038ec3ea5f120494d0ce5a7/cocos/animation/marionette/pose-graph/foundation/pose-graph-node.ts#L30

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/copy-transform.ts:
```diff
@@ -89,7 +89,7 @@ class Workspace {
 
 if (EDITOR) {
-    PoseNodeCopyTransform.prototype.getTitle = function getTitle (this: PoseNodeCopyTransform) {
+    PoseNodeCopyTransform.prototype.getTitle = function getTitle (this: PoseNodeCopyTransform): string | [string, Record<string, string>] | undefined {
```
The same as above.

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/play-motion.ts:
```diff
@@ -148,7 +149,7 @@ class Workspace {
 
 if (EDITOR) {
-    PoseNodePlayMotion.prototype.getTitle = function getTitle (this: PoseNodePlayMotion) {
+    PoseNodePlayMotion.prototype.getTitle = function getTitle (this: PoseNodePlayMotion): string | [string, Record<string, string>] | undefined {
```
The same as above.

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/sample-motion.ts:
```diff
@@ -84,7 +84,7 @@ class SampleMotionWorkspace {
 
 if (EDITOR) {
-    PoseNodeSampleMotion.prototype.getTitle = function getTitle (this: PoseNodeSampleMotion) {
+    PoseNodeSampleMotion.prototype.getTitle = function getTitle (this: PoseNodeSampleMotion): string | [string, Record<string, string>] | undefined {
```
The same as above.

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/set-auxiliary-curve.ts:
```diff
@@ -74,7 +74,7 @@ export class PoseNodeSetAuxiliaryCurve extends PoseNodeModifyPoseBase {
 
 if (EDITOR) {
-    PoseNodeSetAuxiliaryCurve.prototype.getTitle = function getTitle (this: PoseNodeSetAuxiliaryCurve) {
+    PoseNodeSetAuxiliaryCurve.prototype.getTitle = function getTitle (this: PoseNodeSetAuxiliaryCurve): string | [string, Record<string, string>] | undefined {
```
The same as above.

minggo commented on cocos/animation/marionette/pose-graph/pose-nodes/state-machine.ts:
```diff
@@ -72,14 +73,14 @@ export class PoseNodeStateMachine extends PoseNode {
 
 if (EDITOR) {
-    PoseNodeStateMachine.prototype.getTitle = function getTitle (this: PoseNodeStateMachine) {
+    PoseNodeStateMachine.prototype.getTitle = function getTitle (this: PoseNodeStateMachine): string | [string, Record<string, string>] | undefined {
```
The same as above.

PPpro committed:
update review

minggo commented on cocos/asset/asset-manager/deprecated.ts:
```diff
@@ -532,7 +532,9 @@ export class CCLoader {
      */
-    public get md5Pipe () {
+    public get md5Pipe (): {
+        transformURL(url: string): string;
```
The type seems strange.

PPpro replied:
that is because this getter returns an object `md5Pipe`, and `md5Pipe` is an untyped js object, 

so only an automatically inferred type can be generated here

minggo commented on cocos/profiler/counter.ts:
```diff
@@ -58,24 +58,24 @@ export class Counter {
     }
 
-    public alarm () {
+    public alarm (): boolean | 0 | undefined {
```
I think the return type is boolean only.

PPpro replied:
I don't think there's anything wrong with type inference, it's a matter of the method implementation

the return statement is 
```ts
return (
            (this._opts.below && this._value < this._opts.below)
            || (this._opts.over && this._value > this._opts.over)
        )
```

This means that if `this._opts.below` is 0, it may indeed return 0,
or if `this._opts.below` is undefined, it will return undefined

I would consider changing the return type to `boolean | number | undefined`

In this PR, I don't want to modify the method implementation 
to prevent this PR from introducing other hard-to-track issues

PPpro replied:
Interestingly, when we actually start thinking about function return types, we found that the implementation itself has a lot of problems

PPpro replied:
updated https://github.com/cocos/cocos-engine/pull/15314/commits/8843bfe54cd3f8e6f84432948716c8d130ecb233

minggo replied:
Yep, and please create an issue for it.

PPpro replied:
created https://github.com/cocos/cocos-engine/issues/15355

minggo commented on cocos/core/data/decorators/utils.ts:
```diff
@@ -58,7 +58,7 @@ export const emptyDecorator: ClassDecorator & LegacyPropertyDecorator = () => {}
  * 一个忽略所有参数并且返回 `emptyDecorator` 的函数。
  */
-export const emptyDecoratorFn = () => emptyDecorator;
+export const emptyDecoratorFn = (): ClassDecorator & LegacyPropertyDecorator => emptyDecorator;
```
The return type is strange, please double check it.

PPpro replied:
It's OK, `ClassDecorator & LegacyPropertyDecorator` is the type of `emptyDecorator` object

this is an arrow function, and `emptyDecorator` is the return value, not part of the return type

minggo replied:
Why use `&`, i don't understand.

PPpro replied:
`&` means emptyDecorator need to implement both ClassDecorator and LegacyPropertyDecorator these two interface

minggo replied:
But why not `|`?

PPpro replied:
`|` means emptyDecorator is a union type of ClassDecorator and LegacyPropertyDecorator , 
in this case, we need emptyDecorator to implement ClassDecorator and LegacyPropertyDecorator

minggo commented on cocos/core/data/decorators/utils.ts:
```diff
@@ -138,16 +138,17 @@ export function makeEditorClassDecoratorFn<TValue> (propertyName: string): (valu
  */
-export function makeSmartEditorClassDecorator<TValue> (propertyName: string, defaultValue: TValue) {
-    return makeSmartClassDecorator<TValue>((constructor, decoratedValue?: TValue) => {
+export function makeSmartEditorClassDecorator<TValue> (propertyName: string, defaultValue: TValue): ClassDecorator & ((arg?: TValue | undefined) => ClassDecorator)
```
The same as above.

PPpro replied:
it's OK, it's an infered type from the return statement

minggo commented on cocos/core/data/utils/compact-value-type-array.ts:
```diff
@@ -139,11 +142,11 @@ export class CompactValueTypeArray {
 }
 
-function getStorageConstructor (unit: StorageUnit) {
+function getStorageConstructor (unit: StorageUnit): Float32ArrayConstructor | Uint16ArrayConstructor | Uint32ArrayConstructor | Uint8ArrayConstructor | Int8ArrayConstructor | Int16ArrayConstructor | Int32ArrayConstructor | Float64ArrayConstructor {
```
Is there  ArrayConstructor that covers all these types.

PPpro replied:
we do have one called `TypedArrayConstructor`

PPpro replied:
updated https://github.com/cocos/cocos-engine/pull/15314/commits/8843bfe54cd3f8e6f84432948716c8d130ecb233

PPpro committed:
update review
